forge-game/src/main/java/forge/game/zone/MagicStack.java

```diff
@@ -355,21 +355,17 @@ public final void add(SpellAbility sp, SpellAbilityStackInstance si, int id) {
             sp.getActivatingPlayer().addSpellCastThisTurn();
 
             // Add expend mana
-            Map<Player, Integer> players = Maps.newHashMap();
-
-            for(SpellAbility sa : sp.getPayingManaAbilities()) {
-                // This section has some complications that may need to be addressed
-                // How do we differentiate mana paid from Assist
-                // With mana created by opposing effects like Eladamri's Vineyard OR Yurlok of Scorch Thrash
-                Player manaPayer = sa.getActivatingPlayer();
-                if (!players.containsKey(manaPayer)) {
-                    players.put(manaPayer, 0);
-                }
+            Map<Player, Integer> expendPlayers = Maps.newHashMap();
+
+            for (Mana m : sp.getPayingMana()) {
```

Comment:
Yea, maybe Mana just needs to have a FK to Player who "owned" it, that way we can differentiate Vineyard mana from Assist mana.